source/fab/util.py

```diff
@@ -127,18 +147,67 @@ class CompiledFile(object):
     A Fortran or C file which has been compiled.
 
     """
-    def __init__(self, analysed_file, output_fpath: Path):
+    def __init__(self, input_fpath, output_fpath,
+                 source_hash=None, flags_hash=None, module_deps_hashes: Dict[str, int] = None):
         """
-        :param analysed_file:
+        :param input_fpath:
             The file that was compiled.
         :param output_fpath:
-            The object file.
+            The object file that was created.
+
+        """
+        # todo: Should just be the input_fpath, not the whole analysed file
+        self.input_fpath = Path(input_fpath)
+        self.output_fpath = Path(output_fpath)
+
+        self.source_hash = source_hash or 0
+        self.flags_hash = flags_hash or 0
+        self.module_deps_hashes = module_deps_hashes or {}
+
+    #
+    # persistence
+    #
+    @classmethod
+    def field_names(cls):
+        return [
+            'input_fpath', 'output_fpath',
+            'source_hash', 'flags_hash', 'module_deps_hashes',
+        ]
+
+    def to_str_dict(self):
+        """
+        Convert to a dict of strings. For example, when writing to a CsvWriter.
 
         """
-        # todo: A compiled file shouldn't know whether its source file was analysed or not.
-        #       It needn't be in some use cases. This is a code smell and needs revisiting.
-        self.analysed_file = analysed_file
-        self.output_fpath = output_fpath
+        return {
+            "input_fpath": str(self.input_fpath),
+            "output_fpath": str(self.output_fpath),
+            "source_hash": str(self.source_hash),
+            "flags_hash": str(self.flags_hash),
+            "module_deps_hashes": ';'.join([f'{k}={v}' for k, v in self.module_deps_hashes.items()]),
```

Comment:
I don't understand, sorry.

Reply:
Function `field_names` lists all the field names. Then `to_str_dict` repeats all that data again. Then this function repeats it all again. It would be good to have a single source of truth for the field names.

Reply:
I see, thanks. The best solution to that is probably to have a `fields` method which returns `Dict[name, type]`. Because `to_str_dict` and `from_str_dict` would become generic, it would seem well placed in a _"Dictable"_ (or better named) base class. This is quite nice. I can do it here or in a follow-up ticket - but aren't we likely to remove all this with a move to a database sometime soon?